src/core/Akka.Streams/Implementation/Fusing/Ops.cs

```diff
@@ -3756,7 +3757,127 @@ public RecoverWith(Func<Exception, IGraph<SourceShape<TOut>, TMat>> partialFunct
+                    // OnSuccess/OnFailure/OnComplete methods directly. We need to do that via async callbacks.
+                    if (task.IsFaulted) _onFailure(task.Exception);
+                    else if (task.IsCanceled) _onFailure(new TaskCanceledException(task));
+                    else if (task.Result) _onSuccess(enumerator.Current);
+                    else _onComplete();
+                };
+
+                SetHandler(_outlet, this);
+            }
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private void OnComplete() => CompleteStage();
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private void OnFailure(Exception exception) => FailStage(exception);
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private void OnSuccess(T element) => Push(_outlet, element);
+
+            public override void OnPull()
+            {
+                var vtask = _enumerator.MoveNextAsync();
+                if (vtask.IsCompletedSuccessfully)
+                {
+                    // When MoveNextAsync returned immediatelly, we don't need to await.
+                    // We can use fast path instead.
+                    if (vtask.Result)
+                    {
+                        // if result is true, it means we got an element. Push it downstream.
+                        Push(_outlet, _enumerator.Current);
+                    }
+                    else
+                    {
+                        // if result is false, it means enumerator was closed. Complete stage in that case.
+                        CompleteStage();
+                    }
+                }
+                else
+                {
+                    vtask.AsTask().ContinueWith(_handleContinuation);
+                }
+            }
+            public override void OnDownstreamFinish(Exception cause)
+            {
+                var vtask = _enumerator.DisposeAsync();
+                if (vtask.IsCompletedSuccessfully)
+                {
+                    CompleteStage(); // if dispose completed immediately, complete stage directly
+                }
+                else
+                {
+                    // for async disposals use async callback
+                    vtask.GetAwaiter().OnCompleted(_onComplete);
+                }
+                base.OnDownstreamFinish(cause);
+            }
+
+        }
+
+        #endregion
+        private readonly Outlet<T> _outlet = new Outlet<T>("EnumerableSource.out");
+        private readonly Func<IAsyncEnumerable<T>> _factory;
 
+        /// <summary>
+        /// TBD
+        /// </summary>
+        /// <param name="factory">TBD</param>
+        public AsyncEnumerable(Func<IAsyncEnumerable<T>>  factory)
+        {
+            //TODO: when to dispose async enumerable? Should this be a part of ownership of current stage, or should it
```

Comment:
No need to have this here anymore. We're now responsible for disposing the `IAsyncEnumerable` - hence the factory.